Feature Request for Ctrl+A Highlight of Context Menu Descriptions

sagec108 opened the pull request:
Pull request for #304. Added code that allows user to now use Ctrl+A shortcut key to highlight text inside the context menu's description box (even if the description box is not current focus). Ctrl+A will focus the description box, even if the user has not typed a description (highlights nothing).

Also added "Esc" key functionality to close context menu (resolves #611)

manaswis commented:
The ESC functionality works. The CTRL+A doesn't for Mac. Its missing Mac's Command key. You have currently coded only 17 which is 'Ctrl' key for Mac and Windows, however, Select all would be Command + A for Mac. So for coding this, you need to also check the PC the user is on. The same key code might have different meanings in different OSs. See this site - http://keycode.info/. This will give you good info on each press key.

manaswis commented:
Also, see this - https://stackoverflow.com/questions/10655202/detect-multiple-keys-on-single-keypress-event-in-jquery

You need code both keyup and keydown events.

sagec108 commented:
The code I implemented handles both keyup and keydown events on the same function by using a boolean mapping called "down". I added in some code to handle command key detection, but I need a Mac to test it.

jonfroehlich commented:
I brought in my Mac today for you Sage. So you should be able to have a PC
and Mac machine to test code this summer (perhaps Mac can be shared by
multiple users).

Jon

On Mon, Jun 5, 2017 at 10:28 AM, Sage Chen <notifications@github.com> wrote:

> The code I implemented handles both keyup and keydown events on the same
> function by using a boolean mapping called "down". I added in some code to
> handle command key detection, but I need a Mac to test it.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/625#issuecomment-306201909>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABi-9Xij6Y2VrOvuiWS32Bdf1-2TMDkhks5sBBCngaJpZM4NtOb7>
> .
>



-- 
Jon Froehlich
Assistant Professor
Computer Science
University of Maryland, College Park
http://www.cs.umd.edu/~jonf/
@jonfroehlich <https://twitter.com/jonfroehlich> - Twitter

manaswis commented:
It works now! Since you are working on the context menu in this PR, could you make another small change: The mouse pointer turns to the hand symbol (![image](https://cloud.githubusercontent.com/assets/2873216/26835179/aee05b78-4aa5-11e7-8a7d-64370c8170aa.png))  when hovered over the passable and not passable icons on the far left and right sides respectively. This implies that its clickable while its not. It would be super fast fix. I will merge all of these changes together when you are done.

![image](https://cloud.githubusercontent.com/assets/2873216/26834837/84294e04-4aa4-11e7-8862-f03934175b7f.png)

sagec108 commented:
I changed the cursor hovering for the passability image areas to the default cursor, and cursor hovering on the "skip the tutorial" link and the context menu check box to the hand/pointer cursor.

manaswis commented:
@sagec108 I was doing teh final testing with your changes. There is a bug left. So when I select the text using Cmd + A, and then try to cut+paste it, the pasted text is auto-selected each time. Looks like this:

![image](https://user-images.githubusercontent.com/2873216/26949425-ddfe2520-4c67-11e7-8f89-52301a0aa3d1.png)

sagec108 commented:
I think I fixed the bug in my latest commit, I tested on Jon's Mac. Please let me know if you find any other issues.

manaswis commented:
Doesn't quite work yet.

Multiple things are happening:
- When I do Cmd + C then Cmd + V, then it essentially selects the text (isn't visible as it happens really quick) and pastes it over. So I can't get this kind of text using the keyboard:
![image](https://user-images.githubusercontent.com/2873216/27163533-07e88766-5156-11e7-9bac-0c099ddfe5d9.png)

- However, I am able to get the above (pasting multiple times) when I press backspace after the text is selected. BUT this isn't consistent. This doesn't happen every time.

In summary, its broken as its showing erratic behavior.

sagec108 commented:
Did you deselect the text after using Cmd + C? Or did you use Cmd + A -> Cmd + C -> Cmd +V?

If you did the latter, then pasting it over is the natural thing to do (since you never deselect).

manaswis commented:
Yes, it deselects automatically when I paste it. And then if I paste it again, it does the action that I described previously.

So essentially, I did this - Cmd + A -> Cmd + C -> Cmd +V -> Cmd +V -> Cmd +V -- Ideally it should paste multiple times, but it doesn't.

sagec108 commented:
I'm sorry I would have to test this on a Mac to debug it. I'm not quite sure why it is so buggy for Mac, as the functionality seems perfectly fine on Windows.

manaswis commented:
It works! This was a lengthy one to fix. Nice job @sagec108! :)

![image](https://user-images.githubusercontent.com/2873216/27256186-cbf47f12-537b-11e7-8e34-a59b51f782da.png)
